### ISSUE TITLE
Thoughts on management UI?

### ISSUE THREAD

**Issue by dandv** [vsivsi/meteor-job-collection#14]
I'm eager to get a queue management UI working and wanted to check with you for the general architecture and any guidance you might have. Final goal would be to integrate this with Houston, but it can function standalone for now:

- [ ] list queues and jobs
- [ ] show overall progress
- [ ] show progress per job
- [ ] pause/resume/remove job from queue
- [ ] pause entire queue
- [ ] bonus points for further editing, e.g. of the data payload

Our use case is infinitely repeating jobs: ~100 of one type (feed parsing), and several jobs of different types. Makes sense to group by type.

I'm thinking of using [jQuery DataTables](https://github.com/LumaPictures/meteor-jquery-datatables) for its filtering, sorting and editing capabilities, and an admin theme like [SB Admin 2](http://startbootstrap.com/sb-admin-v2).

**Comment by vsivsi** [comment]
Yes, collections need to be declared in both places if you want access on the client.. JobCollection is a collection internally, so goes by the same rules.

**Comment by LarsBuur** [comment]
The main point of my notes to this report got lost (maybe in translation).

What I found strange was that I need to create the JobCollection on both the server and the client - and that I am unable to share a JobCollection on both the server and client using a file in e.g. /collections which should be picked up by both the server and the client.

I am definitely willing to live with this - but it is still strange in my mind.

**Comment by vsivsi** [comment]
@LarsBuur  Thanks for clarifying. Have you looked at the JobCollection sample app?  It has only a single source file for both client and server and they share a single definition of the JobCollection:
https://github.com/vsivsi/meteor-file-job-sample-app/blob/master/sample.coffee#L18

I've not encountered any problems with doing it this way.  If you are seeing an issue, please provide some code in a sample app that replicates the issue and I'd be happy to take a look.

**Comment by vsivsi** [comment]
@LarsBuur Actually I see your steps to reproduce above. Sorry, I lost track of this issue.  I'll take a look and see what I can find.  Thanks.

**Comment by vsivsi** [comment]
So, when I try this using Meteor 1.0.3.1 and job-collection 0.0.18, I am able to partly reproduce what you are describing.  It seems that the reason is three-fold:

1) The `subscribe` on the client should also be inside of a `Meteor.startup()` code block (at least that seems to help...)

2) The `console.log` in the publish function may not print the new job that is saved below it because there's a race condition there.

3) The 1000ms timeout on the `setTimeout` in the client code you provided is often inadequate. Upping that to 10000ms always resulted in the `find` returning a value on my machine. When you restart the server, the client may or may not reload (depending on the reason for the restart), and that will only run when it does. And even when it does, the timeout may expire before the local collection has been synchronized So, the key seems to be actually waiting on the `onReady` callback of the subscribe, rather than picking an arbitrary timeout.  

Another observation is that Meteor seems to be pretty lazy in syncing client subscriptions when there are no reactive cursors (or observers) registered to consume the records, as in this example.

So these changes to your repro code above make it work for me, so long as the client reloads. Also, note that a single JobCollection declaration work for both Server and Client.

```js
var myJobs = new JobCollection('myJobQueue');

///////////////////
// Server

if (Meteor.isServer) {

    myJobs.allow({
        // Grant full permission to anyone
        admin: function (userId, method, params) {
            return (userId ? true : true);
        }
    });

    Meteor.startup(function () {
        // Normal Meteor publish call, the server always
        // controls what each client can see
        Meteor.publish('allJobs', function () {
            var retVal = myJobs.find({});
            return retVal;
        });

        // Create a job:
        var job = myJobs.createJob('sendEmail', // type of job
            // Job data that you define, including anything the job
            // needs to complete. May contain links to files, etc...
            {
                address: 'bozo@clowns.com',
                subject: 'Critical rainbow hair shortage',
                message: 'LOL; JK, KThxBye.'
            }
        );

        // Set some properties of the job and then submit it
        job.priority('normal')
            .retry({
                retries: 5,
                wait: 15 * 60 * 1000
            })  // 15 minutes between attempts
            .delay(60 * 60 * 1000)     // Wait an hour before first try
            .save();

        console.log(JSON.stringify(myJobs.find({}).fetch()));

        // Start the myJobs queue running
        return myJobs.startJobs();
    });
}

///////////////////
// Client
if (Meteor.isClient) {
    Meteor.startup(function () {
        Meteor.subscribe('allJobs', function () {
            console.log("Subscription Ready!");
            console.log(JSON.stringify(myJobs.find().fetch()));
        });
    });
}
```

**Comment by vsivsi** [comment]
One other to note on this issue is that the JobCollection sample application has been fully updated in the past week or so to work with Meter 1.0. Sorry for letting it get out of date, Meteor 1.0 dropped just before the holidays and my long vacation and I didn't have time to update everything before I left.

**Comment by LarsBuur** [comment]
Thank you so much for clarifying this. You example definitely works and we have only one collection defined. Great!

**Comment by KrishnaPG** [comment]
For the admin theme UI, https://github.com/yogiben/meteor-admin fits better instead of starting from [SB Admin2](http://startbootstrap.com/sb-admin-v2).

**Comment by aldeed** [comment]
FYI, we've published a queue management UI package using `aldeed:tabular` here: https://atmospherejs.com/dispatch/tabular-job-collection

It is functional, a work in progress, but we're using it in production.

**Comment by vsivsi** [comment]
@aldeed   Nice!  Thanks for the heads-up!
